### ISSUE TITLE
[SECURITY] This image creates two Admin accounts !

### ISSUE THREAD

**Issue by ChessSpider** [osixia/docker-openldap#161]
Hi!
I noticed unexpected behavior while using your OpenLDAP image.. 

After changing the passwords of `cn=admin,dc=example,dc=org` you can still login using the old password with full rights.  The reason is that this image defines `RootDN` and `RootPW` in `DN: olcDatabase={1}hdb,cn=config` as `cn=admin,dc=example,dc=org`. This is a separate login.

# steps to reproduce
### Start the stuff
```
me@VirtualBox:~/test/docker-openldap/example$ docker-compose up 
```
### Login via PHPLDapAdmin
username:  `cn=admin,dc=example,dc=org`
password: `admin`
### Change the password of `cn=admin,dc=example,dc=org` to `example`
Use hash `{SSHA}bbbIClpMGqvOpa5Yc9x+VS/9PN1czs1B`
Or create a new SSHA hash here: http://projects.marsching.org/weave4j/util/genpassword.php (not for production)

### Login again with `cn=admin,dc=example,dc=org` with the old password `admin`

# Expected behavior:

Login fails

# Observed behavior

Login succeeds, given full root access. 

# Reason why

In  `DN: olcDatabase={1}hdb,cn=config` database the `cn=admin,dc=example,dc=org` is defined as `RootDN` with the initial password as `RootPW`. This is a real root-user with full access that bypasses all ACL.   This password is not changed when you change the password of `cn=admin,dc=example,dc=org`. 

Meaning, when you 

# Impact

If people start the image default password and expect changing

**Comment by osixia** [comment]
Hello,
thanks for reporting this.

Indeed we can add information how to change the admin password. Pull request welcomed.

Maybe adapt your issue title to something less worrying and more accurate.

**Comment by ChessSpider** [comment]
Hi thanks for the response 👍 . I want to emphasize I like and appreciate the work you've done in creating and maintaining this image. 

Any suggestions for title? I will remove the exclamation mark at the end but it is security related and the image by default does create two admins.  

Adding documentation to change the admin password will not fix this vulnerability for existing users of your image.  Also, your image should not create two admin accounts during bootstrap. 

Any current user of the image may be vulnerable. In that light, it might be an option to create a script which removes the `cn=admin,{{ BASE_DN}}` user to mitigate the risk. Or remove the RootDN. Or sync the password of the `cn=admin,{{BASE_DN}}` user to RootDN.  

I don't think there is any good fix for the current users of your image that works best for all your users, and they should be informed so they can decide the best action.  What do you think?

**Comment by osixia** [comment]
Oops sorry,
your title is accurate i read a bit too fast, i just make the full test and your right this is a major issue.

Yes users should be informed, i'm going to add a warning on top of the readme linking to this issue,
maybe we can wait for some user inputs to choose the best option ?

**Comment by ChessSpider** [comment]
Hi thanks for the confirmation. It's a difficult issue. I think most people will use this image and just let it running, without looking here again. Waiting has an obvious risk, but rushing into a decision may cause a lot of damage. 

I suggest this fix for the next update: 

Create a shell script which executes automatically, which:

1. Prints a warning informing them of this issue
2. And either:
2.1 If both accounts exist: Takes the password from `cn=admin,{{BASE_DN}}` and sets it as RootPW in `olcDatabase={1}hdb,cn=config`. Then remove `cn=admin,{{BASE_DN}}`
2.1 If `cn=admin,{{BASE_DN}}` is missing and RootPW of `olcDatabase={1}hdb,cn=config` is admin, remove RootDN and RootPW from `olcDatabase={1}hdb,cn=config`,
2.1 If RootDN/RootPW doesn't exist, do nothing

I think this script should create the situation which people expect.  What do you think?

**Comment by osixia** [comment]
not sure to understand this point: 
"If cn=admin,{{BASE_DN}} is missing and RootPW of olcDatabase={1}hdb,cn=config is admin, remove RootDN and RootPW from olcDatabase={1}hdb,cn=config"

this leads to no admin account ? Except with the config password ?

**Comment by ChessSpider** [comment]
Correct. My reasoning is that, if the user removed the `cn=admin,{{BASE_DN}}` user it does not want to use that admin account. Perhaps it has created a new admin account with different ACLs. 

Of course it is a problem if the user wants to use the RootDN account. However the user could re-add the RootDN account easily.

**Comment by osixia** [comment]
Sounds good to me :)

Just maybe not do this automatically but ask for user consent by adding a boolean environment variable like FIX_SECURITY_ISSUE_161

So when the user start the container with next image version:
- If FIX_SECURITY_ISSUE_161 is undefined or something else than true or false and we are in case 1 or 2 : 
we print a warning message that explain the problem and the fix that will be applied and ask the user to set FIX_SECURITY_ISSUE_161 to true if want to fix this automatically or false to discard this warning
- if FIX_SECURITY_ISSUE_161 is true we apply the fix
- if FIX_SECURITY_ISSUE_161 we do nothing

What do you think?

**Comment by osixia** [comment]
Hello,
Ok to make it autcomatically in a first time.

We will make a release candidate with that and see feedback :)

**Comment by ChessSpider** [comment]
Any progress on the fix? :)

**Comment by osixia** [comment]
we have no time to work on this for now :(
any help would be appreciated

**Comment by lknock** [comment]
These are the same issue, no? https://www.openldap.org/lists/openldap-technical/201212/msg00225.html

From what I understand, there is an authentication identity for root access if you set olcRootDN and olcRootPW. It is not a user. This means you can still create a user with the same DN, which is why it appears there are 2 admins or one admin with two passwords when you do this. 

If olcRootPW is not set, they advise to create the admin user. Otherwise, it is not even necessary. So it's either one or the other then.

**Comment by BertrandGouny** [comment]
This behavior occurs also on a regular debian install of slapd (debian strech, openldap 2.4.44)
i guess this should be reported upstream.

Does anyone have an idea how to fix this ?

**Comment by Bessonov** [comment]
Run in this issue. Is there any way to configure admin password in a secure way? Because `LDAP_ADMIN_PASSWORD` isn't encrypted.

**Comment by BertrandGouny** [comment]
You can set the admin password in a *.startup.yaml environment file, that is deleted right after startup scripts. Or if you use kubernetes run an init container with command line option `-p` and `LDAP_ADMIN_PASSWORD` environment variable that will just bootstrap openldap config and do not run openldap. Then run a regular pod container using the same ldap config and data dirs

**Comment by kopax** [comment]
I am happy I have found this post before. So your conclusion is to remove the `cn=admin,${BASE_DN}` so I do only have one account?

How will I be able to access through phpmyadmin and perform edition to `${BASE_DN}` tree?
